File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/G.Гардероб/solution.py
import sys


def solution(n: int, nums: list[int]) -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(1)
    """
    zeros_idx: int = 0
    twos_idx: int = n - 1
    ones_idx: int = 0
    while ones_idx <= twos_idx:
        if nums[ones_idx] == 0:
            nums[ones_idx], nums[zeros_idx] = nums[zeros_idx], nums[ones_idx]
            if zeros_idx == ones_idx:
                ones_idx += 1
            else:
                zeros_idx += 1
        elif nums[ones_idx] == 2:
            nums[ones_idx], nums[twos_idx] = nums[twos_idx], nums[ones_idx]
            twos_idx -= 1
        else:
            ones_idx += 1

    for idx in range(n):
        if idx:
            sys.stdout.write(' ')
        sys.stdout.write(str(nums[idx]))


if __name__ == '__main__':
    n: int = int(sys.stdin.readline().rstrip())
    nums: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    solution(n, nums)

File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/G.Гардероб/solution.cpp
#include <iostream>
#include <vector>


void solution(int n, std::vector<int>& nums) {
    /*
    Time Complexity O(N)
    Memory Complexity O(1)
    */
    int zerosIdx = 0;
    int onesIdx = 0;
    int twosIdx = n - 1;
    while (onesIdx <= twosIdx) {
        if (nums[onesIdx] == 0) {
            int zerosValue = nums[zerosIdx];
            nums[zerosIdx] = nums[onesIdx];
            nums[onesIdx] = zerosValue;
            if (zerosIdx == onesIdx) {
                ++onesIdx;
            }
            ++zerosIdx;
        } else if (nums[onesIdx] == 2) {
            int twosValue = nums[twosIdx];
            nums[twosIdx] = nums[onesIdx];
            nums[onesIdx] = twosValue;
            --twosIdx;
        } else {
            ++onesIdx;
        }
    }

    for (int idx = 0; idx < n; ++idx) {
        if (idx) {
            std::cout << " ";
        }
        std::cout << nums[idx];
    }
}


int main() {
    int n;
    std::cin >> n;
    std::vector<int> nums;
    for (int i = 0; i < n; ++i) {
        int num;
        std::cin >> num;
        nums.push_back(num);
    }
    solution(n, nums);
}